File: 01-romannumerals/romannumerals.py
numerals = {
  "I": 1,
  "V": 5,
  "X": 10,
  "L": 50,
  "C": 100,
  "D": 500,
  "M": 1000
}


def convert(roman):

	arabic = 0
	counter = 0
	skipNextChar = False

	while counter < len(roman) - 1: 

		if skipNextChar:
			skipNextChar = False
			counter += 1
			continue
		
		thisChar = roman[counter].upper()
		nextChar = roman[counter + 1].upper()

		if numerals[thisChar] < numerals[nextChar]:
			arabic += numerals[nextChar]-numerals[thisChar]
			skipNextChar = True
		else:
			arabic += numerals[thisChar]

		counter += 1

	if not skipNextChar:
		arabic += numerals[roman[-1]]

	return arabic


def main():
	print("Roman to Arabic numeral unit tests:")
	print("III\t3\noutput: " + str(convert("III")))
	print("XII\t12\noutput: " + str(convert("XII")))
	print("VII\t7\noutput: " + str(convert("VII")))
	print("XXVI\t26\noutput: " + str(convert("XXVI")))
	print("IV\t4\noutput: " + str(convert("IV")))
	print("IX\t9\noutput: " + str(convert("IX")))
	print("XL\t40\noutput: " + str(convert("XL")))
	print("LVIII\t58\noutput: " + str(convert("LVIII")))
	print("MCMXCIV \t1994\noutput: " + str(convert("MCMXCIV")))

main()

File: 01-romannumerals/RomanNumerals.java
import java.util.HashMap;

class RomanNumerals {

	public static void main (String[] args)  {
		System.out.println("Roman to Arabic numeral unit tests:");
		System.out.println("III \t3\noutput: " + String.valueOf(romanToInt("III")));
		System.out.println("XII \t12\noutput: " + String.valueOf(romanToInt("XII")));
		System.out.println("VII \t7\noutput: " + String.valueOf(romanToInt("VII")));
		System.out.println("XXVI \t26\noutput: " + String.valueOf(romanToInt("XXVI")));
		System.out.println("IV \t4\noutput: " + String.valueOf(romanToInt("IV")));
		System.out.println("IX \t9\noutput: " + String.valueOf(romanToInt("IX")));
		System.out.println("XL \t40\noutput: " + String.valueOf(romanToInt("XL")));
	}

    public static int romanToInt(String roman) {

    	roman = roman.toUpperCase();

    	HashMap<Character, Integer> numerals = new HashMap<Character, Integer>();
	    numerals.put('I', 1);
	    numerals.put('V', 5);
	    numerals.put('X', 10);
	    numerals.put('L', 50);
	    numerals.put('C', 100);
	    numerals.put('D', 500);
	    numerals.put('M', 1000);

    	int arabic = 0;
		int counter = 0;
		boolean skipNextChar = false;
	
		while (counter < roman.length() - 1) { 
		
			if (skipNextChar) {
				skipNextChar = false;
				continue;
			}
			
			char thisChar = roman.charAt(counter);
			char nextChar = roman.charAt(counter + 1);

			if (numerals.get(thisChar) < numerals.get(nextChar)) {
				arabic += numerals.get(nextChar) - numerals.get(thisChar);
				skipNextChar = true;
			}
			else
				arabic += numerals.get(thisChar);

			counter++;
		}

		if (!skipNextChar)
			arabic += numerals.get(roman.charAt(roman.length() - 1));
			
		return arabic;
	}
}